daffodil-runtime1/src/main/scala/org/apache/daffodil/layers/midBitsToEndTransformer.scala

```diff
@@ -0,0 +1,345 @@
+        argsString.length, argsString.mkString(","))
+    }
+    try {
+      //DFDL is 1-indexed, we are working 0-indexed
+      firstBitIndex0b = argsString(0).toInt - 1
+      numBits = argsString(1).toInt
+    } catch {
+      case _: NumberFormatException => state.SDE(
+        "midBitsToEnd layer transform requires Integer arguements, recieved: %s",
+        argsString.mkString(","))
+    }
+
+  }
+
+  /*
+   * Scala does not seem to support bitwise operations on bytes; and instead silently upcasts to Int.
+   * Spamming toByte is insufficient since negative numbers will get padded with 1 instead of 0
+   * For this reason, we do most of our internal logic in Int, and mask out the upper 24 bits
+   */
+
+  val bitMask: Int = 0x000000ff
+
+  private def doDecode(buff: Array[Byte]): Unit = {
+
+    /*First, move the middle bits to the end.
+     *
+     * Since the layer is byte aligned, we know we are writing them a byte aligned location
+     * We will end up moving a multiple of 8 bits. This is okay, since, after we shift,
+     * any extra bytes will be left passed the length byte, and so will be ignored
+     */
+    var dstByteIndex0b = length
+    var srcByteIndex0b = firstBitByteIndex0b
+    while (dstByteIndex0b < buffSize) {
+      val srcByte = buff(srcByteIndex0b) & bitMask
+      val srcByteNext = buff(srcByteIndex0b + 1) & bitMask
+
+      val byteUpper = (srcByte << firstBitBitIndex0b) & bitMask
+      val byteLower = (srcByteNext >>> (8 - firstBitBitIndex0b))
+      val byte = (byteLower | byteUpper).toByte
+      buff(dstByteIndex0b) = byte
+      dstByteIndex0b += 1
+      srcByteIndex0b += 1
+    }
+
+    /*
+     * When we do the shift, we will end up overwriting all the bits in byte containing firstBitIndex.
+     * We want to preserve all bits from before firstBitIndex
+     */
+    val savedBits = {
+      val srcByte = buff(firstBitByteIndex0b) & bitMask
+      ((srcByte >>> (8 - firstBitBitIndex0b)) << (8 - firstBitBitIndex0b)).toByte
+    }
+
+    /*
+     * Next, perform the shift
+     */
+    dstByteIndex0b = firstBitByteIndex0b
+    while (dstByteIndex0b < length) {
+      val srcByte = buff(dstByteIndex0b + byteShift) & bitMask
+      val srcByteNext = buff(dstByteIndex0b + byteShift + 1) & bitMask
+
+      val byteUpper = srcByte << bitShift & bitMask
+      val byteLower = srcByteNext >>> (8 - bitShift)
+      val byte = (byteLower | byteUpper).toByte
+      buff(dstByteIndex0b) = byte
+      dstByteIndex0b += 1
+    }
+
+    /*
+     * Finally, restore the upper bits of the byte containing firstBitIndex
+     */
+    {
+      val srcByte = buff(firstBitByteIndex0b) & bitMask
+      val byteLower = ((srcByte << firstBitBitIndex0b) & bitMask) >>> firstBitBitIndex0b
+      val byteUpper = savedBits
+      val byte = (byteLower | byteUpper).toByte
+      buff(firstBitByteIndex0b) = byte
+    }
+
+  }
+
+  private def doEncode(buff: Array[Byte]): Unit = {
```

Comment:
At first look, both the doEncode and doDecode function seems overly complex to me, and potentially not optimal. Both do two shifts, one to the right and one to the left. I'm still going through these, but it seems this could be simplified some and elimiate a shift.

Reply:
A lot of the shifting is really just masking. It looks like our Util.Misc has a Bits.maskR and Bits.maskL that would provide a more readable solution.